### PR TITLE
Add 'package' argument to IconData

### PR DESCRIPTION
**Pull request by szakarias**

### CODE REVIEW

**Comment by szakarias** [comment]
cc @xster

**Comment by Hixie** [comment]
Can you elaborate on this? What is this feature used for? When would you set it? (The documentation should include all this information.)

**Comment by Hixie** [comment]
Also this would need tests.

**Comment by Hixie** [comment]
Ah, having seen https://github.com/flutter/flutter/pull/12559 I understand this better now.

I'm surprised that the font name of fonts in subpackages is prefixed by the package name, but I guess that makes sense...

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 1]
```diff
@@ -14,17 +14,24 @@ class IconData {
   ///
   /// Rarely used directly. Instead, consider using one of the predefined icons
   /// like the [Icons] collection.
+  ///
+  /// The [package] argument must be non-null when using a font family that is
+  /// included in a package.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/icon_data.dart]: This sentence should explain why.

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 2]
```diff
@@ -14,17 +14,24 @@ class IconData {
...
-  });
+    String fontFamily,
+    this.package,
+  }) : fontFamily = package == null ? fontFamily : 'packages/$package/$fontFamily';
 
   /// The Unicode code point at which this icon is stored in the icon font.
   final int codePoint;
 
   /// The font family from which the glyph for the [codePoint] will be selected.
   final String fontFamily;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/icon_data.dart]: These docs should mention how to reference font families from subpackages

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 3]
```diff
@@ -14,17 +14,24 @@ class IconData {
...
 
+  // The name of the package from which the font family is included.
+  final String package;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/icon_data.dart]: This should talk about the effect it has on the fontFamily property

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 4]
```diff
@@ -14,17 +14,24 @@ class IconData {
...
+  ///
+  /// The [package] argument must be non-null when using a font family that is
+  /// included in a package.
   const IconData(
     this.codePoint, {
-    this.fontFamily,
-  });
+    String fontFamily,
+    this.package,
+  }) : fontFamily = package == null ? fontFamily : 'packages/$package/$fontFamily';
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/icon_data.dart]: I think we should either have fontFamily be unaffected by package, and then have a separate getter that returns the combined string, or we should have the package argument not set a field. It's weird having fontFamily and package arguments map to fields but not have their values exactly map.

**Comment by Hixie** [comment]
Sorry about all the comments here, my understanding of the patch evolved as I was leaving them.

**Comment by Hixie** [comment]
Ah, I see that I missed the PR that added this to TextStyle.

I think the comments I have here should probably be applied to the TextStyle class.
The IconData consumers can then just pass the package to the TextStyle class without having to know about the special way the font family name is adjusted.

**Comment by xster** [comment]
LGTM % Ian's comments

Thanks for making the change

**Comment by szakarias** [comment]
@Hixie, thanks a lot for the comments! Updated TextStyle in #12577.

**Comment by Hixie** [comment]
LGTM

**Comment by Hixie** [comment]
And thanks @szakarias for addressing the comments!

**Pull request merged**